hidimstat/cpi.py

```diff
@@ -0,0 +1,227 @@
+import numpy as np
+from joblib import Parallel, delayed
+from sklearn.base import (
+    BaseEstimator,
+    TransformerMixin,
+    check_is_fitted,
+    clone,
+)
+from sklearn.metrics import mean_squared_error
+
+
+class CPI(BaseEstimator, TransformerMixin):
+    """
+    Conditional Permutation Importance (CPI) algorithm.
+    :footcite:t:`Chamma_NeurIPS2023` and for group-level see
+    :footcite:t:`Chamma_AAAI2024`.
+
+
+    Parameters
+    ----------
+    estimator: scikit-learn compatible estimator
+        The predictive model.
+    imputation_model: scikit-learn compatible estimator or list of
+    estimators
+        The model(s) used to estimate the covariates. If a single estimator is
+        provided, it will be cloned for each covariate. Otherwise, a list of
+        potentially different estimators can be provided, the length of the
+        list must match the number of covariates.
+    n_perm: int, default=50
+        Number of permutations to perform.
+    groups: dict, default=None
+        Dictionary of groups for the covariates. The keys are the group names
+        and the values are lists of covariate indices.
+    loss: callable, default=mean_squared_error
+        Loss function to evaluate the model performance.
+    score_proba: bool, default=False
+        Whether to use the predict_proba method of the estimator.
+    random_state: int, default=None
+        Random seed for the permutation.
+    n_jobs: int, default=1
+        Number of jobs to run in parallel.
+
+    References
+    ----------
+    .. footbibliography::
+    """
+
+    def __init__(
+        self,
+        estimator,
+        imputation_model,
+        n_perm: int = 50,
+        groups: dict = None,
+        loss: callable = mean_squared_error,
+        score_proba: bool = False,
+        random_state: int = None,
+        n_jobs: int = 1,
+    ):
+
+        check_is_fitted(estimator)
+        self.estimator = estimator
+        self.imputation_model = imputation_model
+        if isinstance(self.imputation_model, list):
+            self.list_imputation_mod = self.imputation_model
+        else:
+            self.list_imputation_mod = []
+        self.n_perm = n_perm
+        self.groups = groups
+        self.random_state = random_state
+        self.loss = loss
+        self.score_proba = score_proba
+        self.n_jobs = n_jobs
+
+        self.rng = np.random.RandomState(random_state)
+
+    def fit(self, X, y):
+        """
+        Fit the covariate estimators to predict each group of covariates from
+        the others.
+        """
+        if self.groups is None:
+            self.nb_groups = X.shape[1]
+            self.groups = {j: [j] for j in range(self.nb_groups)}
+        else:
+            self.nb_groups = len(self.groups)
+        # create a list of covariate estimators for each group if not provided
+        if len(self.list_imputation_mod) == 0:
```

Comment:
I don't like `self.list_imputation_mod`
Maybe `self.list_imputation_models`
bzw, the user is not supposed to manipulate these models ? So this one could be an internal variable,
`self._list_imputation_models` ?